src/server/transaction.cc

```diff
@@ -204,23 +204,25 @@ void Transaction::InitShardData(absl::Span<const PerShardCache> shard_index, siz
   CHECK_EQ(args_.size(), num_args);
 }
 
-void Transaction::RecordMultiLocks(const KeyIndex& key_index) {
-  DCHECK(multi_);
-  DCHECK(!multi_->lock_mode);
-
-  if (multi_->mode == NON_ATOMIC)
-    return;
-
-  auto lock_key = [this](string_view key) { multi_->locks.emplace(KeyLockArgs::GetLockKey(key)); };
+CmdArgList Transaction::CopyMultiKeys(CmdArgList keys) {
+  DCHECK(multi_->mode == LOCK_AHEAD);
+  DCHECK_GT(keys.size(), 0u);
+
+  // Store all unqiue keys in multi data
+  absl::flat_hash_set<std::string_view> seen_keys;
+  for (MutableSlice key : keys) {
+    auto lock_key = KeyLockArgs::GetLockKey(facade::ToSV(key));
+    if (seen_keys.insert(lock_key).second) {
+      multi_->keys.emplace_back(lock_key);
+    }
+  }
 
-  multi_->lock_mode.emplace(LockMode());
-  for (size_t i = key_index.start; i < key_index.end; i += key_index.step)
-    lock_key(ArgS(full_args_, i));
-  if (key_index.bonus)
-    lock_key(ArgS(full_args_, *key_index.bonus));
+  // Replace first 0..n with pointers to stored keys
+  size_t i = 0;
+  for (string& stored_key : multi_->keys)
```

Comment:
1. Are you changing the vector that is passed to this function? 
2. `key[0]` in the caller before the call won't be equal to `key[0]` after the call ?

Reply:
1. Yes and no. I'm changing elements of the span, which is backed by a vector all the way in main_service. It has trivially destructable MutableSlice's so, chaning elements is not a problem
3. Yes, the keys are normalized and made unique, we can return something totally different. But we don't access `keys` anywhere in `StartMultiLockAhead`

Reply:
so after this call main_service keys will hold something different. For example, StartMultiEval keys will change underneath with slice having `[k1, k1, k2, k3] ` it may now have `[k3,k1, k2, k3]` because you internally look at the subset of 3 items but higher levels are not aware.

it's fragile, don't you think? can we maybe enforce passing temprefs? I am trying to think how it can be done without introducing this surprising effect.

Reply:
Actually, once you mention it, we can work up the line and simplify how we check declared keys with EVAL. Will look into it

Reply:
Because every level of indirection stores, filters AND normalized the keys

1. Track EVAL keys
2. Copy multi keys
3. Track keys in lock table

we can remove at least half the work and code